### ISSUE TITLE
check that sbt is on your PATH and that your config is compatible

### ISSUE THREAD

**Issue by fommil** [ensime/ensime-emacs#225]
_From @neverov on October 22, 2014 9:24_

**TL;DR**

* check that `sbt` really is on your `PATH`
* check the variable `ensime-sbt-command` is non-nil (and is valid)
* enable `(setq debug-on-error t)`, try again and open up a ticket with the backtrace (unless the problem becomes obvious), a copy of your `.ensime` (or link to the project), and your operating system details.

---

I'm trying to use `ensime-sbt` plugin, which is defined in `.sbt/0.13/plugins/plugins.sbt` like this: 
```
resolvers += Resolver.sonatypeRepo("snapshots")

addSbtPlugin("org.ensime" % "ensime-sbt" % "0.1.5-SNAPSHOT")
```
I'm generating `.ensime` config using `gen-ensime`. It generates successfully, but when I try to load it in emacs I get the following error: `check that sbt is on your PATH and that your config is compatible with https://github.com/ensime/ensime-server/wiki/Example-Configuration-File [(wrong-type-argument stringp nil)]`. The generated `.ensime` config file is [here](https://github.com/neverov/scala-ensime-test/blob/master/.ensime).

I use GNU Emacs 24.3.1 (x86_64-apple-darwin14.0.0, NS apple-appkit-1343.13).
I created the following tiny project to illustrate the issue: https://github.com/neverov/scala-ensime-test.
My emacs configuration is defined here: https://github.com/neverov/emacs-config (it's located in `~/.emacs.d/` on my machine).

_Copied from original issue: ensime/ensime-server#670_

**Comment by fommil** [comment]
you're using an old version of the sbt-plugin, please update (e.g. by clearing out your ivy cache and rerunning)

**Comment by fommil** [comment]
_From @neverov on October 22, 2014 9:29_

Thanks for your response!
I'm using `0.1.5-SNAPSHOT` as stated in quick start guide [here](https://github.com/ensime/ensime-server/wiki/Quick-Start-Guide#installing-the-ensime-sbt-plugin), which version should I use instead?

**Comment by fommil** [comment]
you've got an old copy of the snapshot, clear out your caches.

**Comment by fommil** [comment]
hmm, actually, it looks like you are current (incidentally, you don't have the java sources installed on your machine so that entry is coming up as `nil`.)

It works for me, so I'm afraid I cannot reproduce. Are you sure you're using the most current version of the `ensime` package from MELPA? Also, turn on debugging as per: https://github.com/ensime/ensime-emacs#ensime-emacs

**Comment by fommil** [comment]
btw, ensime won't be able to see `hi.scala` see #668

**Comment by fommil** [comment]
_From @neverov on October 22, 2014 15:47_

I played a little bit around this for a while, cleaning caches and reinstalling ensime and plugins and still have the issue. I'll try to reinstall everything from scratch (probably, on clean machine) and let you know about the result.
Thank you very much for trying to reproduce the issue!

**Comment by fommil** [comment]
cool, thanks for going through all that. I don't think you need to clean any ivy caches, it looks like the problem is entirely in emacs... if you could turn on backtracing we'll see exactly which line is causing the problem.

**Comment by fommil** [comment]
err, btw, can emacs definitely see the `sbt` binary? Just in the off chance that our warning message is actually correct.

**Comment by fommil** [comment]
_From @neverov on October 22, 2014 17:19_

I have `exec-path-from-shell-initialize` plugin enabled with 
```lisp
(when (memq window-system '(mac ns))
  (exec-path-from-shell-initialize))
```
and just to be sure  
```lisp
(setenv "PATH" (concat "/usr/local/bin/sbt/bin:" (getenv "PATH")))
(setenv "PATH" (concat "/usr/local/bin/scala/bin:" (getenv "PATH")))
```
so, when I run sbt from `shell` or `eshell` inside emacs it starts perfectly.
I'll try to debug and share the results with you.

**Comment by fommil** [comment]
Damn

**Comment by fommil** [comment]
aah! This may be an OS X thing. I noticed that your `java-home` is pointing to the JRE and not the JDK. Try changing it.

**Comment by fommil** [comment]
_From @neverov on October 23, 2014 9:26_

After some debugging I noticed, that while starting ensime server in [`ensime-mode.el`](https://github.com/ensime/ensime-emacs/blob/master/ensime-mode.el#L526) the actual value of `ensime-sbt-command` is `nil`, so you were right about this one. The question is why `executable-find "sbt"` in [`ensime-vars.el`](https://github.com/ensime/ensime-emacs/blob/master/ensime-vars.el#L60) fails. Any ideas?

**Comment by fommil** [comment]
hmm, no idea, I think that is part of stdlib emacs. Maybe it's not compatible with your `exec-path-from-shell-initialize` or maybe you need to make sure that runs before loading ensime.

**Comment by fommil** [comment]
_From @neverov on October 23, 2014 9:57_

Life is pain))
![5448d02e38d3871419ab5b3a](https://cloud.githubusercontent.com/assets/57802/4889468/11f79428-6396-11e4-9fc7-02cdcfe02cbf.png)

**Comment by fommil** [comment]
_From @neverov on October 23, 2014 10:3_

Thanks again for your help!

**Comment by fommil** [comment]
hehe, cool! Well this can serve as an FAQ because it is quite common that sbt is missing from the PATH.

**Comment by fommil** [comment]
_From @AmirKarimi on November 15, 2014 13:50_

@neverov I have the same issue but I couldn't solve it yet. Would you please explain how did you solve it?

**Comment by fommil** [comment]
_From @neverov on November 16, 2014 15:13_

@AmirKarimi in the first place I fixed this by updating the path before evaluating `(require 'ensime)` in my `init.el`. Anyhow, after some last update it's not working again. I'll try to debug this tomorrow and keep you posted.

**Comment by fommil** [comment]
_From @AmirKarimi on November 16, 2014 15:23_

@neverov thanks, it got resolved #693.

**Comment by fommil** [comment]
_From @neverov on November 16, 2014 15:46_

@AmirKarimi awesome! My own problem was simply resolved by updating project and caches)

**Comment by fommil** [comment]
_From @xyalan on January 9, 2015 10:16_

 I have the same issue in gui mode.In terminal mode work well.

**Comment by fommil** [comment]
solutions at the top

**Comment by fommil** [comment]
_From @peterbecich on February 15, 2015 7:32_

I had the same error, and eventually found a different solution.

After updating the Ensime package in Emacs to 20150211.1505, that error appeared.  Deleting `.ensime`, `.ensime_cache`, the Ivy caches in `~/.ivy2` did not fix it.  SBT was on the Path, although the variable `ensime-sbt-command` was nil.  Changing the SBT version used by the project from 0.13.5 to 0.13.7 and regenerating the `.ensime` file replaced the error with 
`could not find or load main class org.ensime.server.Server`.

This went away by just deleting the Ensime package in `*Packages*` and then installing it again.  Now it works again.

**Comment by fommil** [comment]
_From @hedefalk on April 2, 2015 20:43_

I get this same error. I'm just trying to start up Emacs as a reference when working with Atom. I'm in no way an Emacs users so I'm really pretty lost here.

I was using an older Ensime with shell-script for starting the server but as I understand it, it's to be started with sbt now so I need to check how that is done.

I have generated my .ensime with ensime-sbt 0.1.5 and it seems to be right.

My ~/emacs.d/init.el looks like:
```
;; if you're new to the MELPA package manager, just include
;; this entire snippet in your `~/.emacs` file and follow
;; the instructions in the comments.
(require 'package)
(add-to-list 'package-archives
         '("melpa" . "http://melpa.milkbox.net/packages/") t)
(package-initialize)

(when (not package-archive-contents)
  (package-refresh-contents))

;; Restart emacs and do `M-x package-install [RETURN] ensime [RETURN]`
;; To keep up-to-date, do `M-x list-packages [RETURN] U [RETURN] x [RETURN]`

;; If necessary, make sure "sbt" and "scala" are in the PATH environment
(setenv "PATH" (concat "/usr/local/bin/sbt:" (getenv "PATH")))
(setenv "PATH" (concat "/usr/local/bin/scala:" (getenv "PATH")))


(require 'ensime)
(add-hook 'scala-mode-hook 'ensime-scala-mode-hook)
```

Starting ensime I get:

```
Wrote /var/folders/rm/zpyjwkdj6hn1b7gqn1pv3l880000gn/T/ensime_update_17076B0I/build.sbt
Wrote /var/folders/rm/zpyjwkdj6hn1b7gqn1pv3l880000gn/T/ensime_update_17076B0I/project/plugins.sbt
Wrote /var/folders/rm/zpyjwkdj6hn1b7gqn1pv3l880000gn/T/ensime_update_17076B0I/project/build.properties
condition-case: check that sbt is on your PATH and that your config is compatible with http://github.com/ensime/ensime-server/wiki/Example-Configuration-File [(wrong-type-argument stringp nil)]
```

The steps at the top, can I pretty please get "monkey instructions" for performing these steps? I'm really no Emacs user :)

**Comment by fommil** [comment]
hmm. go into `/var/folders/rm/zpyjwkdj6hn1b7gqn1pv3l880000gn/T/ensime_update_17076B0I` and type `sbt update`. that might shed some light on this.

**Comment by fommil** [comment]
re: "monkey instructions" just copy and paste

```
(setq debug-on-error t)
```

into your emacs config file and restart emacs and try again.

**Comment by fommil** [comment]
_From @hedefalk on April 2, 2015 20:49_

Thanks for quick reply, @fommil !

sbt:
```
var/folders/rm/zpyjwkdj6hn1b7gqn1pv3l880000gn/T/ensime_update_17076B0I $ sbt update
Java HotSpot(TM) 64-Bit Server VM warning: ignoring option MaxPermSize=512m; support was removed in 8.0
[info] Loading global plugins from /Users/viktor/Dropbox/Shared settings/sbt/plugins
[info] Updating {file:/Users/viktor/.sbt/0.13/plugins/}global-plugins...
[info] Resolving org.fusesource.jansi#jansi;1.4 ...
[info] Done updating.
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
[info] Loading project definition from /private/var/folders/rm/zpyjwkdj6hn1b7gqn1pv3l880000gn/T/ensime_update_17076B0I/project
[info] Updating {file:/private/var/folders/rm/zpyjwkdj6hn1b7gqn1pv3l880000gn/T/ensime_update_17076B0I/project/}ensime_update_17076b0i-build...
[info] Resolving org.fusesource.jansi#jansi;1.4 ...
[info] Done updating.
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
[info] Set current project to ensime_update_17076b0i (in build file:/private/var/folders/rm/zpyjwkdj6hn1b7gqn1pv3l880000gn/T/ensime_update_17076B0I/)
[info] Updating {file:/private/var/folders/rm/zpyjwkdj6hn1b7gqn1pv3l880000gn/T/ensime_update_17076B0I/}ensime_update_17076b0i...
[info] Resolving jline#jline;2.12 ...
[info] Done updating.
[success] Total time: 6 s, completed Apr 2, 2015 10:48:24 PM
```

**Comment by fommil** [comment]
_From @hedefalk on April 2, 2015 20:51_

emacs:

```
Debugger entered--Lisp error: (error "check that sbt is on your PATH and that your config is compatible with http://github.com/ensime/ensime-server/wiki/Example-Configuration-File [(wrong-type-argument stringp nil)]")
  signal(error ("check that sbt is on your PATH and that your config is compatible with http://github.com/ensime/ensime-server/wiki/Example-Configuration-File [(wrong-type-argument stringp nil)]"))
  error("check that sbt is on your PATH and that your config is compatible with http://github.com/ensime/ensime-server/wiki/Example-Configuration-File [(wrong-type-argument stringp nil)]")
  (condition-case ex (if ensime-auto-generate-config (ensime--maybe-refresh-config nil (quote ensime--maybe-update-and-start) (quote (lambda (reason) (ensime--maybe-update-and-start)))) (ensime--maybe-update-and-start)) ((quote error) (error (format "check that sbt is on your PATH and that your config is compatible with %s [%s]" "http://github.com/ensime/ensime-server/wiki/Example-Configuration-File" ex))))
  ensime()
  call-interactively(ensime record nil)
  command-execute(ensime record)
  execute-extended-command(nil "ensime")
  call-interactively(execute-extended-command nil nil)
```

**Comment by fommil** [comment]
in the buffer called `*scratch*` type

```
(message ensime-sbt-command)
```

and the go to the end of the line and type `C-x C-e` (which executes the lisp its on) and see what gets printed in the minibuffer at the bottom.

**Comment by fommil** [comment]
btw, are you definitely using paulp's sbt (and not some cheap imitation)?

**Comment by fommil** [comment]
_From @hedefalk on April 2, 2015 21:4_

So (message ensime-sbt-command) gives me 'nil', but that's not really news, right? Question is why, no?

Do I need to use paulp's sbt? Did before but it broke somehow…

**Comment by fommil** [comment]
try removing the `PATH` lines. I'm pretty sure /usr/local is on the default path anyway.

**Comment by fommil** [comment]
_From @hedefalk on April 2, 2015 21:13_

I actually added them just before posting, had the same error before. Get the same stack too.

**Comment by fommil** [comment]
doh! we're both eejjits. Of course, PATH is supposed to be the directory not the files themselves. just add /usr/local/bin not the scripts.

**Comment by fommil** [comment]
_From @hedefalk on April 2, 2015 21:25_

Dooh! It's late here in Sweden, I'll blame the clock.

Still the same problem though. Same stack.

Both sbt and Scala are installed via homebrew on Mac. I just "which sbt" and copied it to my init.el. Homebrew uses softlinks, that can't be the problem though can it?

```
$ ll `which sbt`
lrwxr-xr-x  1 viktor  admin    28B Jun 18  2014 /usr/local/bin/sbt@ -> ../Cellar/sbt/0.13.5/bin/sbt


/var/folders/rm/zpyjwkdj6hn1b7gqn1pv3l880000gn/T/ensime_update_17076B0I $ cat /usr/local/Cellar/sbt/0.13.5/bin/sbt
#!/bin/sh
test -f ~/.sbtconfig && . ~/.sbtconfig
exec java -Xms512M -Xmx1536M -Xss1M -XX:+CMSClassUnloadingEnabled ${SBT_OPTS} -jar /usr/local/Cellar/sbt/0.13.5/libexec/sbt-launch.jar "$@"
```

**Comment by lyterk** [comment]
Since this issue is still in the FAQ: 
I normally run the emacs daemon. Moving to a standalone emacs problem resolved the issue for me.
